### PR TITLE
Make handling of non-standard expires_at more tolerant

### PR DESCRIPTION
**Pull request by bbqsrc**
This commit stops token responses from failing to parse when an `expires_at` field is not a long. This is true for Sentry.io's OAuth2 implementation, which responds with an ISO date string for that field, and as this field is not defined in the OAuth 2 RFC, there is no reason for this field to be strictly handled.

### CODE REVIEW

**Comment by iainmcgin** [comment]
I'm a little concerned that this silently discards the expiry time. If the format you are seeing is based on something standard I think I'd rather try parsing that as a fallback, and if neither the integer form nor that form match, throw an exception.

**Comment by iainmcgin** [comment]
I'm inclined to agree with just ignoring expires_at based on your rationale - I've done some digging too and can't remember why we implemented this parsing. It's possible that one of the major IDPs is returning this value, and it would certainly be preferable if everyone did this instead of sending expires_in.

Anyway, based on that, I'm happy to merge your PR as-is, where expires_at is ignored if it doesn't fit our current, largely unfounded, expectations 😆

File: library/java/net/openid/appauth/JsonUtil.java [hunk 1]
```diff
@@ -253,7 +253,9 @@ public static Long getLongIfDefined(
             return null;
         }
 
-        return json.getLong(field);
+        long candidate = json.optLong(field);
+
+        return candidate != 0 ? candidate : null;
```
Comment (iainmcgin) [inline library/java/net/openid/appauth/JsonUtil.java]: Am I right that this might introduce a bug, where the number is _explicitly_ zero in the JSON and this function would then return null? I think you should revert this change if so.

File: library/java/net/openid/appauth/TokenResponse.java [hunk 2]
```diff
@@ -216,7 +216,10 @@ public Builder fromResponseJson(@NonNull JSONObject json) throws JSONException {
             setTokenType(JsonUtil.getString(json, KEY_TOKEN_TYPE));
             setAccessToken(JsonUtil.getStringIfDefined(json, KEY_ACCESS_TOKEN));
             if (json.has(KEY_EXPIRES_AT)) {
-                setAccessTokenExpirationTime(json.getLong(KEY_EXPIRES_AT));
+                long expiresAt = json.optLong(KEY_EXPIRES_AT);
+                if (expiresAt != 0) {
+                    setAccessTokenExpirationTime(expiresAt);
+                }
```
Comment (iainmcgin) [inline library/java/net/openid/appauth/TokenResponse.java]: getLongIfDefined with a null check would more accurately handle this, though I agree that in practice an `expires_at` value of zero isn't likely.

**Comment by bbqsrc** [comment]
I have signed the CLA and the other thing I had to sign.

File: library/java/net/openid/appauth/JsonUtil.java (hunk 1 above)
Comment (bbqsrc) [reply library/java/net/openid/appauth/JsonUtil.java]: The defintion of `optLong` is this:

```java
public long optLong(String name) {
    return optLong(name, 0L);
}
```

This is because `long` is a primitive and cannot be represented as null. Would you like me to try an implementation that uses `Long` instead?

**Comment by bbqsrc** [comment]
Updated with handling that allows a nullable `Long`. Unfortunately this requires catching an exception, however as much checking as possible has been done prior to that case happening.

**Comment by bbqsrc** [comment]
Any chance this can be merged now?

**Comment by iainmcgin** [comment]
Done. I'm _still_ waiting for some CLA-related things on my end with regards to my new employer, so I won't be able to cut a new release with these changes until then. Hold tight ☹️

**Comment by bbqsrc** [comment]
I'm in no rush, just making sure it wasn't forgotten. Thanks! :smile:
